### ISSUE TITLE
Constant Animation Jank in Simple Flutter App

### ISSUE THREAD

**Issue by benweschler** [flutter/flutter#103847]
## Details

I’m observing significant jank on nearly every animation in an app I’m building when run on a physical iPhone 13 Pro Max on iOS 15.4.1 in profile mode. Jank is present throughout the lifetime of the app, including during built-in animations such as scrolling. Oddly, while DevTools does report a handful of slow frames, for the most part it reports frame times under 8ms for animations that clearly display jank. Since I’m observing the issue on iOS, I also built the app using SkSL warm up to rule out shader compilation as a cause — indeed, I still observe jank after running the app using cached shader data.

While the app is not open source, I'm also observing this issue in a skeleton Flutter app, ruling out non-performant dart code as the issue. Similarly, the janking persists when this mini app is built using SkSL warm up. Below is the minimal 37-line dart script I'm using that is reproducing the issue when run in profile mode on the same iPhone 13 Pro Max. Here's a screen recording from that physical device showing janking in basic animations.

https://user-images.githubusercontent.com/91393341/168463816-25603d05-8869-4bb4-8474-7cf19bf1ed83.MP4

Note that the jank I’m observing in the real app visually seems to be more severe with most animations than the janking observed with this minimal script.

While I don’t own a physical Android device to test on, I discussed the issue with another engineer who tested the same 37-line script on a Pixel 5 in profile mode and also noticed an unusual amount of raster jank.

**Target Platforms:** iOS 15.4.1 and possibly Android (unknown version)
**Devices:** iPhone 13 Pro Max and Google Pixel 5

## Minimal Dart Script

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(home: Home());
  }
}

class Home extends StatelessWidget {
  const Home({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: ListView.builder(
        itemBuilder: (_, ndx) => Padding(
          padding: const EdgeInsets.all(15),
          child: Text("Item $ndx", style: const TextStyle(fontSize: 32)),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => Navigator.push(
          context,
          MaterialPageRoute(builder: (_) => const Home()),
        ),
      ),
    );
  }
}
```

## Logs

<details>

  <summary>Flutter Doctor (click to expand)</summary>
  

  ```
  [✓] Flutter (Channel stable, 3.0.0, on macOS 12.3.1 21E258 darwin-x64, locale en-US)
      • Flutter version 3.0.0 at /Users/benjaminweschler/SVDevelopment/flutter
      • Upstream repository https://github.com/flutter/flutter.git
      • Framework revision ee4e09cce0 (5 days ago), 2022-05-09 16:45:18 -0700
      • Engine revision d1b9a6938a
      • Dart version 2.17.0
      • DevTools version 2.12.2
  
  [✓] Android toolchain - develop for Android devices (Android SDK version 32.0.0)
      • Android SDK at /Users/benjaminweschler/Library/Android/sdk
      • Platform android-Tiramisu, build-tools 32.0.0
      • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
      • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
      • All Android licenses accepted.
  
  [✓] Xcode - develop for iOS and macOS (Xcode 13.3.1)
      • Xcode at /Applications/Xcode.app/Contents/Developer
      • CocoaPods version 1.11.2
  
  [✓] Android Studio (version 2021.1)
      • Android Studio at /Applications/Android Studio.app/Contents
      • Flutter plugin can be installed from:
        🔨 https://plugins.jetbrains.com/plugin/9212-flutter
      • Dart plugin can be installed from:
        🔨 https://plugins.jetbrains.com/plugin/6351-dart
      • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
  
  [✓] IntelliJ IDEA Community Edition (version 2021.3.2)
      • IntelliJ at /Applications/IntelliJ IDEA CE.app
      • Flutter plugin version 67.1.3
      • Dart plugin version 213.7371
  
  [✓] Connected device (2 available)
      • Ben’s iPhone (mobile) • 00008110-00140C112E79801E • ios        • iOS 15.4.1 19E258
      • macOS (desktop)       • macos                     • darwin-x64 • macOS 12.3.1 21E258 darwin-x64
  
  [✓] HTTP Host Availability
      • All required HTTP hosts are available
  
  • No issues found!
  ```

</details>

iphone_devtools_output.json: the DevTools output captured during the attached iPhone screen recording.
[iphone_devtools_output.json.zip](https://github.com/flutter/flutter/files/8694561/iphone_devtools_output.json.zip)

pixel_devtools_output.json: the DevTools output from the Pixel 5 sent by the other engineer mentioned earlier.
[pixel_devtools-output.json.zip](https://github.com/flutter/flutter/files/8694563/pixel_devtools-output.json.zip)

**Comment by huycozy** [comment]
Hi @benweschler, I tried to run the demo sample code on my physical devices (iPhone 7, iOS 14.8.1 and Pixel 3a, Android 12). The result will be same as your recorded video. 
It'll be great if you could provide us the detail of reproduction steps as well as the recorded video of the DevTools on your side so that I can verify this issue. Thanks!

**Comment by benweschler** [comment]
Hi @huycozy. Here are the precise steps to reproduce this issue.

### Steps to Reproduce
1. `create` a new Flutter project.
2. Replace any existing dart scripts in the project's source with the minimal dart script posted with my original issue description.
3. `run` the project in profile mode on an iOS device.

As I mentioned in my first post the issue may also be observable on Android, although I haven't verified this myself.

Here's a screen recording of the DevTools timeline when running that minimal script from my original post on an iPhone 13 Pro Max on iOS 15.5. DevTools seems to report some slow frames due to an unexpectedly high amount of raster jank, which caused severe stunting in basic animations. However, the animations were also clearly janking even when DevTools was reporting sub-8 ms frame times.

https://user-images.githubusercontent.com/91393341/168698860-26bddc0f-47c9-4390-b758-c44bf513f436.mov

Here is the DevTools output corresponding to this new recording:
[new_iphone_devtools_output.json.zip](https://github.com/flutter/flutter/files/8703997/new_iphone_devtools_output.json.zip)

**Comment by moffatman** [comment]
It's my experience that you will have this jank when conected to devtools in profile mode, the engine is held up while sending diagnostic data to devtools. I came to this conclusion after wasting a lot of time in diagnosing this phantom lag, it's really only super noticeable at 120fps frame timings. Can you see if you see it happening in release mode as well? I find it's resolved there since there is no communication with devtools.

**Comment by huycozy** [comment]
Hi @benweschler,
Thank you for the detailed reproduction steps. I tried on my side as below demo video. 
The example is run on iPhone 7 (iOS 14.8.1) on profile mode.
There is not much jank but a lot of rasters. Most jank/raster will occur when navigating to new page.

<details>
<summary>Demo</summary>

- First time:

https://user-images.githubusercontent.com/104349824/168732800-415945fd-a725-403b-af90-168128492e39.mp4

- Second time:

https://user-images.githubusercontent.com/104349824/168732818-0883aab3-59bb-4867-9a69-5fc6a400eda0.mp4

</details>

<details>
<summary>flutter doctor -v</summary>

```bash
[✓] Flutter (Channel stable, 3.0.0, on macOS 12.2.1 21D62 darwin-x64, locale en-VN)
    • Flutter version 3.0.0 at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ee4e09cce0 (2 days ago), 2022-05-09 16:45:18 -0700
    • Engine revision d1b9a6938a
    • Dart version 2.17.0
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Android Studio
    • Android Studio at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-1/203.7185775/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] IntelliJ IDEA Community Edition (version 2020.3.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.67.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.40.0

[✓] Connected device (3 available)
    • iPhone 13 (mobile) • 5F5D609C-3B0B-4C48-872C-961928157361 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-2 (simulator)
    • macOS (desktop)    • macos                                • darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 101.0.4951.54

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

```bash
[✓] Flutter (Channel master, 3.1.0-0.0.pre.744, on macOS 12.2.1 21D62 darwin-x64, locale en-VN)
    • Flutter version 3.1.0-0.0.pre.744 at /Users/huynq/Documents/GitHub/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision fd312f1ccf (2 hours ago), 2022-05-16 22:52:09 -0400
    • Engine revision 986700315a
    • Dart version 2.18.0 (build 2.18.0-120.0.dev)
    • DevTools version 2.13.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-32, build-tools 31.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] Android Studio (version 4.1)
    • Android Studio at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/201.7042882/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[!] Android Studio
    • Android Studio at /Users/huynq/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-1/203.7185775/Android Studio Preview.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] IntelliJ IDEA Community Edition (version 2020.3.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.67.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.40.0

[✓] Connected device (4 available)
    • iPhone (mobile)    • d9a94afe2b649fef56ba0bfeb052f0f2a7dae95e • ios            • iOS 14.8.1 18H107
    • iPhone 13 (mobile) • 5F5D609C-3B0B-4C48-872C-961928157361     • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-2 (simulator)
    • macOS (desktop)    • macos                                    • darwin-x64     • macOS 12.2.1 21D62 darwin-x64
    • Chrome (web)       • chrome                                   • web-javascript • Google Chrome 101.0.4951.64

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.

```
</details>

**Comment by rotorgames** [comment]
Looks like duplicate https://github.com/flutter/flutter/issues/101653

**Comment by Hixie** [comment]
Looking at the pixel trace, the slowest frames just have slow rasters for no apparent reason...

**Comment by Hixie** [comment]
I am told the problem does not go away when Text nodes are replaced with ColoredBox nodes.

**Comment by dnfield** [comment]
CAn you please include a timeline trace? I'm unable to load the attached file, but inspecting it closely it does not appear to be a full timeline but instead just frame data. It'd be helpful to see a full trace.

It'd also be helpful to see a trace with `--trace-skia` enabled.

**Comment by zoocityboy** [comment]
Do you have any reason why you don't use the keys, extent and itemCount?

**Comment by b3nni97** [comment]
I have the same problem especially with complex animations or lists these lags are stronger. These lags are strangely not displayed in the profiler.

**Comment by iosephmagno** [comment]
Hello! We do experience these lags on our app too. Code is closed but I can confirm this is serious issue.

Also, can we agree that performance issues are at least P3 + “urgent” label?

**Comment by luckysmg** [comment]
Hi all. I noticed this issue. And I noticed that you devices is IPhone 13 Pro Max and you are dragging on the screen to test the ListView and dragging during navigation stack. I think may be this issue can be solved by https://github.com/flutter/engine/pull/35592

**Comment by b3nni97** [comment]
For me it's the same with 60hz devices...

**Comment by prithvee** [comment]
I'm facing the same issue, tested on :

- iPhone 11:  Consistent jank, Flutter performance tab fps jumps below 30.
- iPad 7th gen: Works much better, no noticeable jank, fps stays at 60.
- OnePlus 7 Pro: Lags initially then recovers (unlike the iPhone which never recovers).

The issue on the iPhone sometimes goes away when the app is sent to background and then brought back to foreground. Since my app has draggable widgets and InteractiveViewer, the jank is highly noticeable when compared to scrolling, making it almost unusable. 

Also verified this behavior with other apps from the Appstore built with flutter and the issue still exists in those apps.

**Comment by iosephmagno** [comment]
We also have this issue with 60hrz devices

**Comment by luckysmg** [comment]
Tested on iPhone12, with a simple `ListView` in **Profile mode.**

Notice the constant junk. But the frames reported in browser ‘s seems not match that in Android Studio's timeline
Here is the preview video. We can see when there are lots of **red frames** in Android studio's timeline, but in browser seems not so much... I don't know if it it a bug of devtool......

https://user-images.githubusercontent.com/49340347/189286968-7622faaf-0f94-4698-979d-874949ceeff4.mp4

**Comment by iosephmagno** [comment]
We too confirm lots of red frames, scrolling is laggy/janky on both ios and android. On ios things are very bad though.

**Comment by dnfield** [comment]
@kenzieschmoll - https://github.com/flutter/flutter/issues/103847#issuecomment-1241560625 seems like a possible devtools bug compraed to the intellij plugin?

**Comment by luckysmg** [comment]
> It's my experience that you will have this jank when conected to devtools in profile mode, the engine is held up while sending diagnostic data to devtools. I came to this conclusion after wasting a lot of time in diagnosing this phantom lag, it's really only super noticeable at 120fps frame timings. Can you see if you see it happening in release mode as well? I find it's resolved there since there is no communication with devtools.

I found that too......In release mode I can't see very transparent junk with my eyes^_^

**Comment by prithvee** [comment]
> > It's my experience that you will have this jank when conected to devtools in profile mode, the engine is held up while sending diagnostic data to devtools. I came to this conclusion after wasting a lot of time in diagnosing this phantom lag, it's really only super noticeable at 120fps frame timings. Can you see if you see it happening in release mode as well? I find it's resolved there since there is no communication with devtools.
> 
> I found that too......In release mode I can't see very transparent junk with my eyes^_^

Not only in release version of my app, this issue exists even in other apps that have been already published on the App Store, eg. Hamilton, Flutter Europe (Tested on iPhone 11, iOS 15.6.1).

**Comment by iosephmagno** [comment]
Yes guys, we confirm lags and overall janky animation while scrolling on release mode. Since you dont see the issue on your devices, can you please share videos? A simple list of items comprising a small pic and a couple of texts (title/description) is already enough to trigger the issue.

We also experience a lag at screen first opening animation, which doesnt resolve with shaders warmup. This lag will occur again when app is moved from background to foreground.

**Comment by luckysmg** [comment]
When app is moved from bg to fg, Seems like this https://github.com/flutter/flutter/issues/96381#issuecomment-1186215653

**Comment by iosephmagno** [comment]
@luckysmg yes it might be that. I think the original sin here is to consider these severe issues as p4, not fixing them for months (or years). Steve Jobs would have dismantled the shack here.

Also, clients are asked to fix these core issues themselves mentioning the open source nature of the project. 99% of clients have no time/resource for that, and in some cases no knowledge for that (eg. Skia related issues).

**Comment by dnfield** [comment]
I believe this bug is trying to deal with at least four separate issues:

1. A relatively simple list view application shows severe jank, more evidently on iOS.
2. There seems to be worse jank on resume from background on iOS.
3. DevTools might be inaccurate in it's reporting, either under or over reporting jank (??)
4. People who are not happy with the prioritization scheme used by this issue tracker.

The second item has an issue filed. The first and third items require a either an application that consistently reproduces the issue or a full trace with --trace-skia that can be shared (no one has provided this yet).  Ideally someone would try this as a systrace type of trace in profile or release mode as well to rule out issues Dart has with reading the system clock in an expensive way on iOS.

The fourth issue should not be brought up again in this thread (including, but not limited to, posts of people saying "me too" which adds nothing but distraction for an engineer trying to work on this topic).  @iosephmagno in particular, this is not the first time I have seen you try to use an unrelated issue to campaign for changes to how flutter prioritizes things. 

Unfortunately, this issue in it's current state isn't actionable. Doing either of the things I've requested would make it actionable.

**Comment by Hixie** [comment]
@iosephmagno Despite our conversation by private e-mail, which included a detailed conversation about our prioritisation schema, you continue to act in an entitled fashion. I'm afraid this is a violation of our code of conduct and I have applied a 30-day block on your account.

**Comment by roybeitner** [comment]
+1
Having severe lag on scroll mostly when resuming from background or launching the app.
The lag is much more noticeable with light colored widgets (not sure if it helps in any way), lag is reduced when using my specific app in dark mode.
Tested on iPhone XS and iPhone 13 Pro Max.
Android on the other hand is pretty smooth

**Comment by chooyan-eng** [comment]
Same issue here with my iPhone 11 (iOS 16.0).

What I tried and found are listed below.

1. Ran a simple app with ListView by `flutter run --profile --trace-skia -d [device_id]`, referencing ["Shader compilation jank" page on flutter.dev](https://docs.flutter.dev/perf/shader#what-is-shader-compilation-jank), but I didn't find `GrGLProgramBuilder::finalize` appearing. (screenshot is attatched below)
2. Tried "SkSL warmup" by `futter run --profile --bundle-sksl-path flutter_01.sksl.json -d [device_id]` after collecting `flutter_01.sksl.json` file with `--cache-sksl --purge-persistent-cache` option, but nothing changed.
3. Tried to check systrace by `futter run --profile --trace-systrace -d [device_id]`, but app crashed right after launching with the message `Oops; flutter has exited unexpectedly: "failed to connect to http://127.0.0.1:60799/".`.

<img width="1489" alt="image" src="https://user-images.githubusercontent.com/20849526/192946932-4be33f38-9bd2-4900-823f-4e2e4c729265.png">

My `flutter --doctor` is showing below.

```
$ flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.2, on macOS 12.5.1 21G83 darwin-arm, locale ja-JP)
[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.0.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.61.2)
[✓] Connected device (3 available)
[✓] HTTP Host Availability
```

<details>
  <summary>Code to reproduce</summary>
  
```dart
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(primarySwatch: Colors.blue),
      home: _MyHomePage(),
    );
  }
}

class _MyHomePage extends StatelessWidget {
  /// icons for each items in ListView
  static final icons = [
    Icons.pedal_bike,
    Icons.javascript,
    Icons.eco,
    Icons.zoom_in_rounded,
    Icons.qr_code,
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Scroll Junk Test')),
      body: ListView.builder(
        itemCount: 100,
        itemBuilder: (context, index) {
          return Padding(
            padding: const EdgeInsets.all(16),
            child: Row(
              children: [
                Text('Item No. $index'),
                const Spacer(),
                Icon(icons[index % 5]),
              ],
            ),
          );
        },
      ),
    );
  }
}
```
</details>

**Comment by girish54321** [comment]
Hi, @chooyan-eng I am all so facing same iuess have you found any fix for this 

https://github.com/flutter/flutter/issues/104376

**Comment by girish54321** [comment]
I can confirm the even in very simple app is slow and janky

Note app is ruing in release mode not debug 

Video attached. https://drive.google.com/drive/folders/15j5oBu6KbSDYQOSCjWKQ35Y8Z6nzmY2a?usp=sharing

iPhone XR : OS : iOS 16.0

**Comment by marcinFDT** [comment]
iPhone 14 Pro iOS 16.02 has the same problem

**Comment by girish54321** [comment]
Not only new app but all apps form App stores apps are slow and janky 
may be something happened in iOS 15 update

**Comment by b3nni97** [comment]
I've been noticing this problem for some months now, but I've never been able to create a simple reproducible example where these lags occur permanently.

**Is it possible that this has anything to do with GC? The lags seem to appear randomly, sometimes scrolling is buttery smooth and in between lags keep appearing which are not visible in the profiler.**

I have had this problem since version 3.0 and it saddens me to see how little attention is paid to this problem. Especially when compared directly to Native iOS apps, the Flutter apps feel like "cheap" copies.

Is there a way to publish an issue that can only be reproduced with more complex apps and not with a simple example?

**Comment by Hixie** [comment]
GC would appear in the trace, as would shader compilation.
The step-like rendering of the Rasterizer::DoDraw call means that that thread got descheduled while it was running.
Usually the way to find the problem is to do a system-level trace and figure out exactly what was happening on the device at the time of that frame. I know that's possible on Android, not sure if it's possible on iOS.

**Comment by dnfield** [comment]
A systrace may help show if we're getting ahead of the GPU too much.

**Comment by dnfield** [comment]
Actually in this case, you'd probably want a metal system trace through instruments. But yes, that really long frame that's descheduled in your screenshot is almost certainly something where we're waiting on the GPU.

**Comment by Hixie** [comment]
@b3nni97 I recommend filing a new bug for your case.

**Comment by b3nni97** [comment]
I have now investigated a little further and have noticed a few interesting things.
The traces and video recordings are from an iPhone X.

To ensure that I was able to record multiple traces and that they remain somewhat comparable, I adjusted the example a little bit. 
Now it automatically scrolls down on startup and then scrolls up.
In addition, the example also includes the bool field "useHeavyWidgets" which, when turned on, runs a 
loop, so that each build takes longer. (According to Performance Overlay this should not cause any lags, but in Systrace several missing vsyncs are visible).

**Systrace 1: useHeavyWidgets = false**
Here you can see a lag after about 4 seconds, which is not visible in the performance overlay.
![lag-systrace1](https://user-images.githubusercontent.com/18357332/196296653-67d4efcc-b90f-41e3-9994-747544abfe19.png)

Video recording: (Since screen recordings do not record the full 60 FPS the micro-lags are not visible, but you can see that there are no frames over 16ms in the performance overlay)
https://drive.google.com/file/d/1T4MHb1KeSOa67uPnbnPHlIGvBt6fJLCs/view?usp=sharing

Full Trace:
https://drive.google.com/file/d/13Qcp6q52ocH1yDX54q-aUJTStGUY6L7D/view?usp=sharing

**Systrace 2: useHeavyWidgets = true**
Here you can see the first lag after about 3.5 seconds, the second at about 3.7 seconds and the third after 4.5 seconds.
As with Systrace 1, these lags are not visible according to the performance overlay.
![lag-systrace2](https://user-images.githubusercontent.com/18357332/196298520-8f902157-8e20-4821-9553-f64c8b2ad904.png)

Video recording: (again no spikes above 16ms)
https://drive.google.com/file/d/1NnTMyRHwU5iS_e9hZmFsFgv9_guC79tk/view?usp=sharing

Full Trace:
https://drive.google.com/file/d/1sYbm7NC7oFqSoM9o2MfD-W4IsdohT7za/view?usp=sharing

**I'm not sure if these missing frames in the traces really has anything to do with the lags in our apps. I hope the Flutter team can do something with this data.**

**Comment by akihisasengoku** [comment]
we are also facing the same problem.

- Poor performance when scrolling ListView. 30-40fps
- Reproduce immediately after launching the app
- There are devices that reproduce and devices that do not reproduce

iPhone | OS | Reproducing
-- | -- | --
iPhone SE | 14.6 | 
iPhone SE3 | 15.6 |
iPhone X | 13.3.1 |
iPhone 11 | 16.0 | ✅
iPhone 11 | 15.4.1 | ✅
iPhone XR | 15.6.1 |
iPhone 14 Pro | 16.0 |
iPhone 13 mini | 16.0.2 |

This is a serious issue for us and we hope it will be resolved.

**Comment by Boydroidnl** [comment]
I'm seeing this issue to on an iPhone Xs. It is awful and prevents me from getting my client to consider switching to Flutter. It makes them uncomfortable that this kind of bugs stay unsolved for so long.

It is easily to be reproduced using the Flutter showcase app from GSkinner https://flutter.gskinner.com/wonderous/

Building the Wonderous app on 3.3.4 myself gives the same problem (release version).

The ticket https://github.com/flutter/flutter/issues/104376 shows a perfect video on this issue, but i just got closed and linked to this one.

**Comment by Boydroidnl** [comment]
@TommasoFramba I'm tagging you because I pointed to your issue in the previous comment.

I tried the Wonderous app with Flutter 3.3.4 and I haven't been able to reproduce it (didn't test it thoroughly, but it was very easy to reproduce).

Anyone else here seeing this is fixed on 3.3.4?

**Comment by b3nni97** [comment]
@Boydroidnl  This PR only helps with Android. For me the problem is reproducible with the latest master on an iPhone X and 12 Pro Max

**Comment by YuriyGrishin** [comment]
I have the same problem on iphone 11, iphone x, iphone 12.

**Comment by guid-empty** [comment]
And yes, we too. Current issue is still block us to release new version.

**Comment by nexron171** [comment]
Same on iPhone XS Max iOS 15.6 and iPhone 13 Pro iOS 16.1.
In case of XS Max fps is about 20
In case of 13 Pro fps is about 60

If open task manager (to see app in task manager preview) fps is normal 60 and 120 respectively. Also enabling screen recording has same effect.

This bug is unstable - sometime it captures at first start, sometime at second or moving from bg to fg constantly.

code from video looks like
`
void main() {
  runApp(TestApp());
}

class TestApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Test'),
        ),
        body: ListView.builder(
          itemBuilder: (context, row) =>
              row == 0 ? CircularProgressIndicator() : Text('$row'),
          itemCount: 99,
        ),
      ),
    );
  }
}
`

I can reproduce it on any Flutter version 2.5+ even on 3.3.6

https://user-images.githubusercontent.com/6318346/199231941-dba42353-db40-424a-b934-ada970e52374.mp4

**Comment by nexron171** [comment]
<img width="1309" alt="изображение" src="https://user-images.githubusercontent.com/6318346/199232430-e8071f40-ecfa-42a3-92dd-702140e2cdcf.png">

**Comment by fzyzcjy** [comment]
(Someone invited me to have a look at this issue so here are my two cents)

So my guess is, maybe we should know a bit more what exactly is slow. For example, from the last profile above, it seems to be GPUSurfaceMetal::Submit. So if I were to debug the problem, maybe we can modify the https://github.com/flutter/engine C++ code, and add a few more timeline events inside that submit call. Then we will clearly see which *exact* one line of code is super slow. But I guess it may even be more than one bug, since another profile reveals slowness in Rasterizer::DoDraw and as Hixie suggests it can be thread scheduling problem...

**Comment by JonLatane** [comment]
I've been experiencing this issue in implementing [a small-scale social network app](https://github.com/JonLatane/jonline/tree/main/frontend), but have found/implemented a workaround for my situation. Essentially, if you build/run the app, I've added a setting for a ["startup sequence"](https://github.com/JonLatane/jonline/blob/776fc6aa3ce3d6ad8557e7e7d56189978c107221/frontend/lib/screens/home_page.dart#L170-L225). (If you don't feel like building you could try the [Flutter Web build](https://jonline.io/app) just to see what it does, though of course the jank patterns don't happen this way on web, and on web the sequence is disabled by default.) 

It just goes through all the tabs programmatically on app startup, and scrolls the most intensive of the app's scroll views. But importantly: If you toggle off the startup sequence setting in the app's settings, you get the jank at startup on iOS. If you leave it on, you'll see some jank *during* the startup sequence (which could be hidden by changing opacity), but it goes away after launch. This approach eliminates the jank within 3-5s as opposed to 30+s or sometimes *minutes* using the app manually, and at least makes for a better UX until this is fixed.

**Comment by b3nni97** [comment]
@JonLatane But this is about a different issue.
What you describe helps against Shader Compilation Jank and there is also a solution for a long time: https://docs.flutter.dev/perf/shader

@nexron171 As mentioned above, there are several causes of Jank, which were unfortunately thrown together here. You mean Jank that appears when you resume the app from the background. This can also be easily reproduced if you drag the Control Center down and up, for example.
For this there is already another issue #96381

The OP means Janks that occur when you see "simple" animation, like scrolling a list or swiping back from CupertinoPageRoutes.

**Comment by JonLatane** [comment]
@b3nni97 The issue I note in my app is not shader jank. SkSL warmup does not fix it. The solution I linked does. It only happens on first startup, not when resuming the app from the background (unless it was already happening). My app is experiencing this issue, not shader jank, as far as I can tell. And the solution, while not that pretty, fixes it when SkSL warmup does not.

**Comment by b3nni97** [comment]
@JonLatane If it's not a shader jank, then create a new issue with a reproducible example create and a before and after video.

IMHO the example from the OP has nothing to do with your problem.

**Comment by JonLatane** [comment]
@b3nni97 I literally linked a reproducible example. It is the same issue that's reported here. If I created another issue, that would just mean this issue is duplicated. What I have linked here is literally an example of this issue, that can be easily turned on and off without having to reinstall the app. I hope it will help in debugging the issue.

**Comment by achillelrc** [comment]
This issue discredits our apps, this is incredible that sometimes the engine runs smoothly af and others it looks like a badly optimized cordova app...

**Comment by Hixie** [comment]
@achillelrc If you have a case that's reproducible, it would be great to file an issue with the code and a trace so that we can examine it more closely. Start [with the performance issue template](https://github.com/flutter/flutter/issues/new?assignees=&labels=created+via+performance+template&template=5_performance_speed.md&title=) and make sure to include all the requested details.

**Comment by Boydroidnl** [comment]
@Hixie but there are a lot of examples mentioned above to reproduce this issue. I even had it with the Wonderous app directly downloaded from the Play Store or building it myself.

**Comment by delfme** [comment]
> we are also facing the same problem.
> 
> * Poor performance when scrolling ListView. 30-40fps
> * Reproduce immediately after launching the app
> * There are devices that reproduce and devices that do not reproduce
> 
> iPhone	OS	Reproducing
> iPhone SE	14.6	
> iPhone SE3	15.6	
> iPhone X	13.3.1	
> iPhone 11	16.0	✅
> iPhone 11	15.4.1	✅
> iPhone XR	15.6.1	
> iPhone 14 Pro	16.0	
> iPhone 13 mini	16.0.2	
> This is a serious issue for us and we hope it will be resolved.


I can reproduce it also on iphone 13 pro. Janks always occur after app launch, then autoresolve after a few taps and then occur back at every app resume. Although thread focuses on iOS, I can reproduce it on some android models as well (eg. Xiaomi mi lite). 
Issue is not related to shaders compilation janks (checked this multiple time) and occur even with lightweight or empty screens.

I’m testing with flutter 3.3.7 stable and ios 15.6.1.

**Comment by Hixie** [comment]
> @Hixie but there are a lot of examples mentioned above to reproduce this issue.

The problem is that there are lots of separate issues. Each one needs to be investigated separately. To do that we need to have distinct bugs filed for each one. There isn't a single "jank" issue to fix.


> I even had it with the Wonderous app directly downloaded from the Play Store or building it myself.

The current jank issues in the Android version of the Wonderous app should go away when we migrate it to the new graphics backend (Impeller), they are (as far as I know) all related to shader compilation (and so go away if you keep using the app). The Wonderous app was made as a demo of what Impeller could do on iOS, we mostly just published the Android version because we could. :-)

**Comment by Hixie** [comment]
@delfme if you are seeing jank that you are sure is definitely not shader-related but that only occurs on app startup and then resolves itself, in release builds (not debug), please do file a bug with a reproducible test case because that sounds like a new problem we haven't heard of before and we should definitely like to investigate further. Start [with the performance issue template](https://github.com/flutter/flutter/issues/new?assignees=&labels=created+via+performance+template&template=5_performance_speed.md&title=) and make sure to include all the requested details.

**Comment by b3nni97** [comment]
@Hixie Should I also create an extra issue? 
I have attached the systraces above, but unfortunately we are not making any progress.

**Comment by Hixie** [comment]
@b3nni97 yes please. comments in this bug are likely to get buried in the noise. feel free to reach out on the discord too (#hackers-performance channel) once you have filed an issue with a reproducible test case if you want help trying to pin down the cause.

**Comment by Boydroidnl** [comment]
> 
> > I even had it with the Wonderous app directly downloaded from the Play Store or building it myself.
> 
> The current jank issues in the Android version of the Wonderous app should go away when we migrate it to the new graphics backend (Impeller); they are (as far as I know) all related to shader compilation (and so go away if you keep using the app). The Wonderous app was made as a demo of what Impeller could do on iOS, we mostly just published the Android version because we could. :-)

Sorry for the mix-up! The Android version is working like a charm! I meant the App Store instead of the Play Store.
Because it is using the Impeller backend, I was mostly surprised that this issue was visible there.

I'll follow your advice and create a performance issue later. Thanks!

**Comment by Hixie** [comment]
If you're seeing jank with Wonderous on iOS please definitely file bugs and include as much detail about your setup as possible — down to things like what language settings you're using, what apps are running in the background, what kind of network you have, definitely what device you have, etc etc etc, there's probably some configuration that we haven't taken into account.

**Comment by samarsheikh001** [comment]
Facing horrible jank in ios 11, flutter engine 3+. I am about to lose my job. Please fix this issue. We don't need any new features.

**Comment by jari-kuipers** [comment]
@samarsheikh001 If you are experiencing jank of such quantities that you're about to lose your job, then you might be facing some different issue regarding UI blocking code. That should be fixable in your own code.

Jank like this is not a reason for anyone to lose their job.

**Comment by samarsheikh001** [comment]
> @samarsheikh001 If you are experiencing jank of such quantities that you're about to lose your job, then you might be facing some different issue regarding UI blocking code. That should be fixable in your own code.
> 
> Jank like this is not a reason for anyone to lose their job.

I would love to put blame on myself, if I were facing constant lag issues. It's so random as discussed above. Reproduced in iPhone 11. But iPhone 14 pro and 13 I couldn't reproduce it. Sometimes it is smooth sometimes it's not.

**Comment by delfme** [comment]
I will also try to provide more data but I hope flutter team has already a good understanding of it.

I also hope we dont have to wait for Impeller. It will take much time for it to become production-ready (6months delay already paid here)

**Comment by moffatman** [comment]
Yeah I suspect some issues in frame scheduling / pacing. It could be interesting to try to accept 2-frame latency if we can be sure that we won't miss a frame. It's my understanding that if (ui+raster) goes from 15ms to 17ms, you will miss a frame, even if both ui and raster on their own are well under 16ms. And I don't think that will be shown in the profiler. Especially noticeable at 120fps.

**Comment by achillelrc** [comment]
Same status, sometimes it's butter smooth, going out and in the app and it is very janky.
Particularly noticeable on a bouncing scroll when scrolling to the top (sliver header is even worse).
No differences between the two cases on devtools (ui+raster < 10ms)
Reproduced on my physical iPhone 14 pro.

**Comment by delfme** [comment]
> @delfme I have the same problem with Impeller.
> 
> Maybe it's a scheduling problem that in some moments only 58 frames are sent to the platform and Flutter thinks it got every VSYNC (that's why no spikes in the profiler). With the systrace I could reproduce this issue a few times.

Yes let me also confirm that issue occurs with Impeller as well. 

Tested with flutter 3.3.7 stable on Iphone 13 pro and ios 16.1.1

**Comment by delfme** [comment]
Tested 2 sample codes above and issue is reproducible on iphone 13 pro, flutter 3.3.7, ios 16.1.1. May I know why these samples are not enough? Issue is such that I don’t think opening new issues with a similar title and code would be more helpful. Did I miss something?

**Comment by delfme** [comment]
If that is helpful for flutter team, I can create a 2-3 screens app that reproduces:
- janks while scrolling list/gridview
- janky screen animations

As reported here in #103847 , #96381 , #104376 (because #31513 was not a final fix) there are a few issues by which we have these janks and micro-janks (reproducible on latest stable 3.3.9). They occur when app is already opened and when app resumes from background. Flutter is amazing but currently no ios app can be shipped and we should try our best to fix this situation. Im also talking to the community, it makes no sense that we keep coding something we cannot ship.

**Comment by prithvee** [comment]
The network searching indicator animation, mentioned here : https://github.com/flutter/flutter/issues/115036#issuecomment-1368147419 , is most probably the root cause of the iOS jank issues (even on Impeller). I can confirm the workarounds mentioned in this comment work because I have been testing my app on an iPhone 11 with no sim card in it. My iPad with no sim support does not have the jank issue because there is no network searching animation. Hope the devs can look into this specifically 🤞

**Comment by achillelrc** [comment]
Not able to reproduce on 3.7 update, Iphone 14 pro ios 16.1.
Maybe related to GC, not sure what really changed but very cool many thanks

**Comment by luckysmg** [comment]
> Not able to reproduce on 3.7 update, Iphone 14 pro ios 16.1. Maybe related to GC, not sure what really changed but very cool many thanks

Not sure... But I can also reproduce this on master channel.. IPhone 13 Pro iOS 16.2

**Comment by luckysmg** [comment]
Also, about the issue that Devtools timeline histogram not accurate, has been fixed by
- https://github.com/flutter/engine/pull/38674 

So junk frame can be transparently seen on flutter master channel using devtools.


<img src="https://user-images.githubusercontent.com/49340347/215752593-320a5d54-ca25-41e4-b25a-94a1240bdd61.png" width=400 />

**Comment by b3nni97** [comment]
I can also reproduce the problem on the latest master.
Tested on an iPhone 14 Pro Max and iPhone 12 Pro Max.

**Comment by b3nni97** [comment]
I can no longer detect any lags on an iPhone 14 Pro on the latest master, everything runs super smooth.
I believe #39359 has fixed this problem. Big thanks @luckysmg

**Comment by luckysmg** [comment]
Yes.Should be much smoother than before.😄

**Comment by achillelrc** [comment]
I was still able to reproduced even after my (early) victory claim and I can confirm that I am not able to reproduce the issue on the master channel, thank you @luckysmg !

**Comment by luckysmg** [comment]
I think This has been fixed by linked PR. Very happy that we finally solve this issue and make flutter much smoother than before. The fix is in cp request and maybe very soon to be in next stable and beta release.

**Comment by jmagman** [comment]
https://github.com/flutter/engine/pull/39359 is available in Flutter stable 3.7.2 as well as the master channel.  
It has not yet been hotfixed to the beta channel, pending request at https://github.com/flutter/flutter/issues/120139.
